libcperciva/util/sock_util.c

```diff
@@ -276,3 +278,74 @@ sock_addr_prettyprint(const struct sock_addr * sa)
 		return (strdup("Unknown address"));
 	}
 }
+
+/**
+ * sock_util_ensure_port(addr):
+ * Allocate a new string to serve as the address for sock_resolve.
+ * If ${addr} contains a port number or is the address of a Unix domain
+ * socket, duplicate that string; if not, add a port number of ":0".
+ */
+char *
+sock_util_ensure_port(const char * addr)
+{
+	char * bind_addr;
+	char * a;
+	char * b;
+	char * c;
+
+	/* Sanity check. */
+	assert(addr != NULL);
+
+	/* Get info about $addr. */
+	a = strchr(addr, ':');
+	b = strrchr(addr, ':');
+	c = strrchr(addr, ']');
+
+	/* Figure out what type of address $addr is. */
+	if (addr[0] == '/') {
+		/* If the address starts with '/', it's a Unix domain socket. */
+		return (strdup(addr));
+	} else if (a == NULL) {
+		/* If there is no ':', it's an ipv4 addr that needs a ":0". */
+		if (asprintf(&bind_addr, "%s:0", addr) == -1) {
+			warnp("asprintf");
+			goto err0;
+		}
+	} else if (a == b) {
+		/*
+		 * If there's a single ':', then it's an ipv4 addr which already
+		 * contains a port.
+		 */
+		return (strdup(addr));
+	} else {
+		/* If there are two different ':', it's ipv6 address. */
+
+		/* If there is no ']', we need to add "[]" and ":0". */
```

Comment:
No no no.  The user MUST place IP addresses inside `[]`.  If they screw that up, they get what they get.  If we say "well we know what the user meant" here then we just train users to get it wrong in other places too.